### PR TITLE
[SPARK-27953][SQL] Save default constraint with Column into table properties when create Hive table

### PR DESCRIPTION
**Pull request by beliefer**
## What changes were proposed in this pull request?
**Background**
Default constraint with column is ANSI standard.
Hive 3.0+ has supported default constraint ref:https://issues.apache.org/jira/browse/HIVE-18726
But Spark SQL implement this feature not yet.

**Design**
Hive is widely used in production environments and is the standard in the field of big data in fact. 
But Hive exists many version used in production and the feature between each version are different.

Spark SQL need to implement default constraint, but there are three points to pay attention to in design:
First, Spark SQL should reduce coupling with Hive. 
Second, default constraint could compatible with different versions of Hive.
Thrid, Which expression of default constraint should Spark SQL support? I think should support `literal`, `current_date()`, `current_timestamp()`. Maybe other expression should also supported, like `Cast(1 as float)`, `1 + 2` and so on.

We want to save the metadata of default constraint into properties of Hive table, and then we restore metadata from the properties after client gets newest metadata.The implement is the same as other metadata (e.g. partition,bucket,statistics).

Because default constraint is part of column, so I think could reuse the metadata of StructField. The default constraint will cached by metadata of StructField.

**Detail of this PR**
This is a sub task to implement default constraint.
This PR will solve the issue that save default constraint into properties of Hive table or data source table.

There exists some issue in this PR:
First, how to check a number specified by somebody compliance with the accuracy and scope of the data type, like float, double.
Second, some code looks not very elegant, I hope to improve it with your suggestions.

**Brother PR**
This PR is related to [https://github.com/apache/spark/pull/24372](24372). If this PR finish, unselected target column can be inserted into the default value, while running `insert into`.
After this PR,  I will continue open other PR about default constraint, like alter table, desc table.

## How was this patch tested?

UT

### CODE REVIEW

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
Retest this please.

**Review by gatorsmile — COMMENTED** [review]
@beliefer Thank you for initiating this. This is not a small work. Could you have a design doc? We need to investigate the impact of DEFAULT on all the other DDL/DML commands and the impact on the data source APIs. 

Personally, I think we might need to create an umbrella JIRA and estimate the sizing.

**Comment by beliefer** [comment]
@gatorsmile Thanks for your review. As you said, this is not a small work. I created a parent jira `SPARK-27943` and described the design simply. I created four sub jira used to each task. If I find other sub task, I will add new sub jira.

**Comment by beliefer** [comment]
@wangyum Help me review this PR, thanks!

**Comment by srowen** [comment]
I don't feel confident enough to review changes to the SQL language support here

**Comment by beliefer** [comment]
> I don't feel confident enough to review changes to the SQL language support here

It doesn't matter, thanks.

**Comment by gatorsmile** [comment]
@beliefer Before submitting PRs, could we first start it with a design doc? Ping me if the design doc is ready to review. Thanks!

**Comment by beliefer** [comment]
> @beliefer Before submitting PRs, could we first start it with a design doc? Ping me if the design doc is ready to review. Thanks!

@gatorsmile I get it. But I want what format is required to design the document? If the design doc is  ready, how I pass it to you?

**Comment by beliefer** [comment]
> @beliefer Before submitting PRs, could we first start it with a design doc? Ping me if the design doc is ready to review. Thanks!

@gatorsmile Thanks for your reply. The design doc is ready, how I pass it to you?

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 1]
```diff
@@ -735,7 +735,7 @@ colTypeList
     ;
 
 colType
-    : identifier dataType (COMMENT STRING)?
+    : identifier dataType (COMMENT STRING)? (DEFAULT defaultExpression=expression)?
```
Comment (lipzhu) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Is that `defaultExpression=expression` scope too big for DDL default constraint? In my memory, the common default constraint are `NULL`, `NUMBER`, `STRING`, `CURRENT_DATE`, `CURRENT_TIMESTAMP`.

Comment (beliefer) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: > Is that `defaultExpression=expression` scope too big for DDL default constraint? In my memory, the common default constraint are `NULL`, `NUMBER`, `STRING`, `CURRENT_DATE`, `CURRENT_TIMESTAMP`.

Thanks for your review. As your said, The description of this PR contains a discussion about the scope of default constraint. Do we need to implement other expressions, like `Cast(1 as float)`, `1 + 2` and so on ?

Comment (lipzhu) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: This is Oracle's default constraint. https://docs.oracle.com/javadb/10.8.3.0/ref/rrefsqlj30540.html#rrefsqlj30540__sqlj64478

Comment (beliefer) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: @lipzhu It's worth to reference, but we need to look at the actual situation on Spark SQL.

Comment (beliefer) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: @lipzhu I reduced the scope of default constraint.  Thanks.

**Comment by dongjoon-hyun** [comment]
Hi, @beliefer . For the umbrella issue, the subtask JIRA ID is enough.

**Comment by beliefer** [comment]
> Hi, @beliefer . For the umbrella issue, the subtask JIRA ID is enough for the title.

OK. Thanks for your reminder.

**Comment by beliefer** [comment]
@gatorsmile  The design doc of default constraint is ready.

**Comment by LiShuMing** [comment]
> @gatorsmile The design doc of default constraint is ready.

what's the progress of this pr?  As https://issues.apache.org/jira/browse/SPARK-29119 also associate with this pr.

I think this will be a useful function for users to handle default value or computed columns;

 @beliefer You can put your design doc on the Google Docs for more details and add comparisons with other engines, eg:
- Oracle: https://docs.oracle.com/javadb/10.8.3.0/ref/rrefsqlj30540.html#rrefsqlj30540__sqlj64478
- Postgres: https://issues.apache.org/jira/browse/SPARK-29119
- Mysql: https://dev.mysql.com/doc/refman/5.6/en/constraints.html
- Cockroach: https://www.cockroachlabs.com/docs/stable/default-value.html

**Pull request closed without merging**